### ISSUE TITLE
Smart wallet middleware

### ISSUE THREAD

**Issue by michaelfig** [Agoric/Agoric#4398]
## What is the Problem Being Solved?

The existing Wallet UI is very close to using a model that is possible to implement via CosmJS-based middleware directly to the on-chain backend.  This issue is to track the pieces needed for an end-to-end.

## Description of the Design

On-chain, per-account state is accessible via Cosmos queries, events, and simple marshalled eventual sends can be made via transactions.

- [x] on-chain notifiers are published by middleware both as updates to queriable state, and susbcribable Cosmos events containing notifier records serialised in the context of the account.  This is possible with the `x/swingset` storage API.
- [x] UI middleware subscribes to events, then issues initialisation queries to obtain and deserialise those records.
- [ ] replace ocaps with a Presence that submits a `MsgWalletAction` Cosmos transaction to do a semantic operation (i.e. only one message needed per UI action)
- [ ] these actions are powerful, so may need confirmation from an external signer (Keplr or MetaMask)
- [ ] an "authz" designation from that signer to a password-protected, per-wallet-UI key can cut down on confirmations #4406 
- [ ] any state changes will be published via the notifiers, so that multiple client UIs can be kept consistent
- [ ] the async result of a `MsgWalletAction` is also published as a Cosmos event, but is designed only to report success or failure to the (single) calling UI initiating it (if any)

NOTE: the wallet backend implementation is specifically not included in this item.  For now, the backend will be a stub that produces:

```ts
type WalletBackend = Notifier<{
  contacts: Notifier<{
    meta: {
      id: string,
    },
    name: string,
    lastReceivedMessage: any,
    actions: ERef<{
      sendMessage: (data: any) => void,
    }>,
  }>,
}>

E(contact.actions).sendMessage('Hello, world!');
```
 
## Security Considerations

Key management is important.  Will need to be reviewed by our security experts.

## Test Plan

Integration testing, probably manual for now.

**Comment by dckc** [comment]
@turadg to split out read-write as a separate issue

**Comment by dckc** [comment]
I'd like to keep using the checklist here, so I'm re-opening this.

In particular, our initial choice of marshaller for the wallet state publishes purses etc. so it needs work, as discussed in https://github.com/Agoric/agoric-sdk/pull/5701#discussion_r917077689

**Comment by dckc** [comment]
tldr: It looks like the `walletFactory` (aka multi-tennant wallet) should register for `BRIDGE.WALLET` events, look up the relevant wallet by address, and forward actions there.

Some of the "actions" and "schema" stuff is still fuzzy to me; here's hoping it becomes clear as we code it up.

## Detail

We have keplr `offlineSigner` integration in https://github.com/Agoric/agoric-sdk/blob/master/packages/wallet/ui/src/util/SuggestChain.js
(another example: https://github.com/Agoric/faucet/blob/cosmjs-rpc/src/gift.js )

@samsiegart asked: how do I map something like `E(purse).deposit(...)` into a payload like in `client.signAndBroadcast(fromAddress, payload, Agoric.fee);`?

We know address. `payload` is a serialized protobuf thing; Looks like  `WalletAction` / `WalletSpendAction` are the relevant ones. 
https://github.com/Agoric/agoric-sdk/blob/21eb21d43390ba51c4d2bc5baecb1b684ee66c52/golang/cosmos/proto/agoric/swingset/msgs.proto#L14-L17

That raises a question of the `action` field:
https://github.com/Agoric/agoric-sdk/blob/21eb21d43390ba51c4d2bc5baecb1b684ee66c52/golang/cosmos/proto/agoric/swingset/msgs.proto#L60-L61


on receipt, the go code seems to do:
https://github.com/Agoric/agoric-sdk/blob/2f9eb8b7364bc1bcfab0399d443dee98e66f47b2/golang/cosmos/x/swingset/keeper/msg_server.go#L99-L103

then I think we get to:
https://github.com/Agoric/agoric-sdk/blob/98d35809ec28b26aa3854eedd1e1af2c9801efbe/packages/cosmic-swingset/src/block-manager.js#L104-L105

**Comment by dckc** [comment]
## marshaling public + private stuff

@samsiegart I broke ground on the problem, starting with a failing test:

https://github.com/Agoric/agoric-sdk/blob/dc-smartwallet-marshal/packages/wallet/api/test/test-union-marshal.js b4db2c4 
_pls excuse lack of branch discipline just now; it's stacked on unrelated work_

### background

The design sketch above includes:
> - on-chain notifiers are published by middleware both as updates to queriable state, and susbcribable Cosmos events containing notifier records serialised in the context of the account.

The notifier records include things like brands that are also published by the AMM, and we want to recognize them as equal.

See also Jan 21 discussion
https://github.com/Agoric/agoric-sdk/issues/3756#issuecomment-1019018331

**Comment by dckc** [comment]
Currently we have `MsgWalletAction` and `MsgWalletSpendAction` with a single `action` field (aside from owner); I wonder how this relates to the `json`, `trivial`, `zoe`, `capdata`, and `unsafe`  from an internal `wallet.proto` I just found.

The related **Wallet Legibility** doc has an example: `E(purse.actions).deposit(myPayment)` becomes `['applyMethod', receiver, purse.actions, 'deposit', [myPayment]]`.

It looks like `receiver` and `purse.actions` are supposed to collapse, and `myPayment` is actually an amount:

```js
marshalData1 = JSON.stringify({
  body: [
    'applyMethod',
    {"@qclass":"slot", index: 0},
    'deposit',
    [
      {
        amount: {“@qclass”:”bigint”, digits: “3000000”},
        brand: {"@qclass": "slot", index:1 },
      }
    ]
  ],
  slots: [[‘purse’, ["MyPurse"]], [‘brand’, ["IST"], {decimalPlaces:6}]],
});
```

**Comment by michaelfig** [comment]
wallet.proto is the evolution and replacement of WalletAction and
WalletSpendAction.

-- 
Michael FIG
Software Engineer, Agoric

**Comment by dckc** [comment]
## road not taken: needs pipelining / shortening

one of @michaelfig 's ideas turns out to be a bridge too far.

See also https://github.com/Agoric/agoric-sdk/discussions/5933

**Comment by dckc** [comment]
I started asking @erights about a puzzle in this space, and he asked for a diagram.

_got started using http://www.webgraphviz.com/ but it's time for a meeting... so this is WIP_

```graphviz

digraph g{
  rankdir=TB;
  subgraph cluster_dapp_amm {
    label="dApp\norigin=amm.agoric.app"
    ammBoot
  }
  ammBoot -> bridgeBoot
  subgraph cluster_wallet_bridge {
    label="wallet bridge\norigin=wallet.agoric.app"
    bridgeBoot -> localStorage
  }
  subgraph cluster_keplr {
    label="keplr extension"
  }
  subgraph cluster_wallet_ui {
    label="wallet UI\norigin=wallet.agoric.app"
    localStorage -> X
  }
  subgraph cluster_smart_wallet {
    label="smartWallet"
    X -> smartWallet
  }
}
```

**Comment by dckc** [comment]
As noted in #3901 , we found the continuing invitation pattern sufficient for voting and for submitting oracle price updates. I'm not inclined to track this issue further. If there's anything important left in here, here's hoping we can scope it more clearly.
